components/bl-credit-card-component/src/card-form.js

```diff
@@ -0,0 +1,160 @@
+import { useMemo } from 'react';
+
+import {
+  formatCreditCardNumber, formatCVC, formatExpirationDate, getExpirationDate, validateCardDetails,
+} from './helpers';
+
+export function CardForm(props) {
+  const { cardNumber, cardholderName, expiry, cvc, cvcVisibility, setFocus } = props;
+  const {
+    submitButtonLabel, cardNumberFieldPlaceholder, cardholderNameFieldPlaceholder,
+    expiryFieldPlaceholder, cvcFieldPlaceholder, labelsVisibility,
+  } = props.component;
+
+  const {
+    handleSubmit, handleNumberChange, handleCardholderNameChange,
+    handleExpiryChange, handleCVCChange, onCardFocus, onCardBlur,
+  } = useFormActions(props);
+
+  const formFields = useMemo(() => [
+    {
+      label      : 'Card Number:',
+      type       : 'tel',
+      name       : 'number',
+      placeholder: cardNumberFieldPlaceholder,
+      pattern    : '[\\d| ]{16,22}',
+      value      : cardNumber,
+      onChange   : handleNumberChange,
+    },
+    {
+      label      : 'Name on card:',
+      type       : 'text',
+      name       : 'name',
+      placeholder: cardholderNameFieldPlaceholder,
+      pattern    : '[a-z A-Z-]+',
+      value      : cardholderName,
+      onChange   : handleCardholderNameChange,
+    },
+    {
+      label      : 'Expiration Date:',
+      type       : 'tel',
+      name       : 'expiry',
+      placeholder: expiryFieldPlaceholder,
+      pattern    : '\\d\\d / \\d\\d',
+      value      : expiry,
+      onChange   : handleExpiryChange,
+    },
+    {
+      label      : 'CVC:',
+      type       : cvcVisibility ? 'tel' : 'password',
+      name       : 'cvc',
+      placeholder: cvcFieldPlaceholder,
+      pattern    : '\\d{3,4}',
+      value      : cvc,
+      onChange   : handleCVCChange,
+    },
+  ], [
+    cardNumber, cardNumberFieldPlaceholder, cardholderName, cardholderNameFieldPlaceholder, cvc, cvcFieldPlaceholder,
+    cvcVisibility, expiry, expiryFieldPlaceholder, handleCVCChange, handleCardholderNameChange, handleExpiryChange,
+    handleNumberChange,
+  ]);
```

Comment:
We could move this logic into hook - useFormFields.
Also, I assume it should be separated memo for each field. For example changed only cvc - but we recalculate all fields.